### PR TITLE
feat(rl): Cortex-serverless dispatch via arctic_platform.rl (zero integration change)

### PR DESCRIPTION
**Pull request by sfc-gh-kganesan**
Depends on / stacked on top of [#47](https://github.com/Snowflake-AI-Research/Arctic-Platform/pull/47). **Do not merge before #47** — base branch is `mwyatt/unified-client-3` so the diff is only the compat + dispatch + env-override delta.

Tracked by [#49](https://github.com/Snowflake-AI-Research/Arctic-Platform/issues/49).

## Why

Both merged upstream integrations —
[NovaSky-AI/SkyRL#1837](https://github.com/NovaSky-AI/SkyRL/pull/1837) (`integrations/arctic_rl/`) and [verl-project/verl#6422](https://github.com/verl-project/verl/pull/6422) (Arctic `RemoteBackend` in `arctic_platform/integrations/verl/adapter.py`) — construct their client via one call:

```python
from arctic_platform.rl import ArcticRLClientConfig, create_arctic_rl_client
client = create_arctic_rl_client(config, server_state)
```

**Both hardcode `backend="local"` in that construction** (SkyRL config.py line 570, verl adapter.py line 559), so setting `backend=cortex` in yaml is a silent no-op today. This PR delivers Cortex serverless enablement without touching either adapter — the launcher exports env vars and `create_arctic_rl_client` rewrites the config before dispatch.

### Compat surface on `arctic_platform/client/*`

Every call site both integrations reach for is now available on the unified client + Cortex transport.

- `arctic_platform/client/config.py` — `extra="ignore"` (was `"forbid"`) so legacy fields pass through. `_apply_legacy_aliases` collapses `backend={local, dss-platform, neutrino}` → `{onprem, cortex}` (with `DeprecationWarning`), aliases `sample_gpus` → `sampling_gpus`, drops `log_prob_engine` / `sampling_engine` / `reference_model` / `job_name` / `experiment_name`. Adds on-prem parity fields (`log_prob_ds_config`, `ds_worker_config`, `arctic_inference_config`, `full_determinism`) threaded to the on-prem transports via `_init_payload`.
- `arctic_platform/client/client.py` — `training_job_id` / `sampling_job_id` / `log_prob_job_id` / `get_server_state()` properties. `fwd_bwd` / `fwd_no_grad` grow `**legacy_kwargs` folded into the transport body (SkyRL: `post_processors=[...]`; verl: `reference_model=…`, `context=…`, `router_replay=…`). `sync_weights(cuda_ipc=False, low_memory=False)`. All colocation-lifecycle ops (`wake_ / sleep_ × {training, inference, log_prob}`, `empty_training_cache`, `weight_norm`, `save_weights`). `_flatten_metrics` bubbles `metrics.grad_norm` / `.loss` to the top level and mirrors `loss` ↔ `avg_loss`.
- `arctic_platform/client/transport.py` — adds nine colocation-lifecycle ops to the canonical `OPS` set.
- `arctic_platform/client/transports/cortex.py` — `_fwd_no_grad` and `_log_probs` handlers; colocation ops registered as `_colo_noop`; `_normalize_train_body` translates the verl-GRPO `{batch, meta, processing}` envelope into Cortex's RPC style; `_shape_train_response` shims loss-only Cortex responses into the on-prem envelope and aliases `model_outputs` → `batch` so `response["batch"]["log_probs"]` works uniformly.
- `arctic_platform/client/transports/onprem_ray.py` — accepts an optional `server_state=` parameter so verl's driver → Ray-worker reconnect flow can reattach across processes.

### `arctic_platform.rl` → Cortex dispatch

- `arctic_platform/rl/config.py` — `backend: Literal["local", "dss-platform"]` widens to include `"cortex"`. Cortex-only fields added: `cortex_host` / `_database` / `_schema` / `_endpoint` / `_pat_env_var` / `_base_url`, `max_seq_len`. `_derive_host_port` short-circuits on cortex, and now also when the caller passed both `host` and `port` explicitly (avoids an eager ray_cluster → tensordict import chain on CPU-only test envs).
- `arctic_platform/rl/client.py` — `create_arctic_rl_client` early-branches when `config.backend == "cortex"` and returns via `_cortex_dispatch.build_cortex_client`. `ArcticRLHTTPClient` / `ArcticRLRayClient` moved to lazy imports inside the on-prem branch so the Cortex code path never triggers the on-prem server chain.
- `arctic_platform/rl/_cortex_dispatch.py` (new): `_to_unified_config` translates the legacy config to unified shape; `_CortexClientShim` wraps `arctic_platform.client.ArcticRLClient` and re-exposes the legacy async surface both integrations reach for.

### Env-var Cortex override — bridges the adapter-hardcode gap

`create_arctic_rl_client` calls `_maybe_override_from_env(config)` before dispatch. When `ARCTIC_RL_BACKEND=cortex` is set, the incoming (hardcoded-local) config gets rewritten to cortex with fields sourced from the env. This is what makes "zero integration change" actually work today, given the adapter hardcodes.

Recognized env vars:

- `ARCTIC_RL_BACKEND=cortex` — toggle; any other value is a no-op.
- `CORTEX_BASE_URL`, `CORTEX_HOST` — target.
- `CORTEX_DATABASE`, `CORTEX_SCHEMA`, `CORTEX_ENDPOINT`, `CORTEX_PAT_ENV_VAR`.
- `CORTEX_MAX_SEQ_LEN` — int; bad values logged and dropped.

Explicit fields on the input config win over env — env is a fallback for adapters whose yaml doesn't yet thread cortex knobs.

### Lazy `arctic_platform/rl/__init__.py`

Eager exports (pydantic only): `ArcticRLClientConfig`, `WeightSyncConfig`. Everything else (`create_arctic_rl_client`, `WeightSyncCoordinator`, all `processors.*` re-exports) loads via PEP 562 `__getattr__` on first attribute access. Result: a Cortex-only driver needs `arctic-platform` + `pydantic` + `requests`. Nothing else. Pinned by `test_cortex_path_never_imports_onprem_transports` + `test_onprem_transports_are_lazy_at_module_level`.

## Tests

**116 pass locally on a driver with only `pydantic` + `requests` + `fastapi` + `uvicorn` + `safetensors` + `torch` installed. No vllm, no ray, no arctic_inference, no tensordict.**

Breakdown:

- `tests/client/test_skyrl_verl_compat.py` — 26 tests pinning every call site SkyRL + verl make against the unified client (using a `_RecordingTransport` shaped like the on-prem server).
- `tests/client/test_client_ops.py` — updated for the expanded `sync_weights` body and the nine new lifecycle ops.
- `tests/client/test_rl_cortex_dispatch.py` — 29 tests total:
  - `TestConfigTranslation` (5), `TestShimAsyncSurface` (10), `TestShimSyncSurface` (3), `TestPropertySurface` (2), `TestFactoryDispatch` (3).
  - `TestEnvOverride` (6, new): env unset is no-op; env flips local → cortex; every recognized env field threaded; explicit config fields win; non-"cortex" toggle values ignored; bad `CORTEX_MAX_SEQ_LEN` warned and dropped.
- `tests/e2e/test_cortex_transport_smoke.py` (new, 18 tests) — full plumbing E2E: driver → unified client → CortexTransport → real HTTP → `fake_cortex_gs` → back:
  - `TestInitialize` — CreateJob + WaitForJob polling completes; sub-jobs captured.
  - `TestCoreOps` (8) — every op the transport dispatches (`fwd-bwd`, `fwd-no-grad`, `step`, `save-checkpoint`, `generate`, `log-probs`, `sync-weights`, `reset-prefix-cache`) round-trips through real DSSST1-chunked octet upload + response decode. `fwd-no-grad` returns a real logprobs tensor at `response["batch"]["logprobs"]` of shape `(batch, seq_len)` after `model_outputs → batch` alias.
  - `TestColocationNoops` (9) — every colocation lifecycle op registered under `_colo_noop` returns `{}` cleanly.

## The fake Cortex GS

`tests/e2e/fake_cortex_gs.py` speaks every REST route the transport actually hits (`CreateJob` / `GetJob` / `forward-backward` / `forward-no-grad` / `generate` / `step` / `save` / `log-probs` / `operation` / `cancel` / `GET requests/{id}`), decodes DSSST1 chunked uploads via `wire.decode_byte_chunks`, and returns shape-plausible canned responses. Fake for training values (random losses, random logprobs) but **plumbing-honest**: same wire codec, same octet-stream framing, same response envelopes. Doubles as an executable reference for what the real Cortex-training endpoint needs to accept.

Runnable standalone: `python -m tests.e2e.fake_cortex_gs --port 8080`. Test suite uses the in-thread `serve_in_background()` helper.

## User-facing UX (what a launcher does to go serverless)

**Zero code change on either integration.** The launcher exports env vars:

```bash
export ARCTIC_RL_BACKEND=cortex
export CORTEX_BASE_URL=https://cortex.snowflakecomputing.com   # or a mock
export CORTEX_DATABASE=my_db
export CORTEX_SCHEMA=rl
export CORTEX_ENDPOINT=cortex-training
export CORTEX_PAT_ENV_VAR=CORTEX_PAT   # PAT itself in CORTEX_PAT
```

Then `run_gsm8k_grpo_arl.sh` (verl) or `run_gsm8k_grpo_4gpu.sh` (SkyRL) runs unchanged; both adapters still build a `backend="local"` config but the factory rewrites it before dispatch.

## What this validates

- Full client-side chain end-to-end, no mocks at any layer between the driver and the REST endpoint. The 18 smoke tests hit the fake GS via real HTTP over a real socket — no `requests-mock`, no `respx`. Wire codec, chunked upload, response polling, response shim, all exercised.
- Both integrations reach the Cortex path via the env-var override with zero adapter change.
- The on-prem path is byte-identical to `main` and stays lazy; `test_cortex_path_never_imports_onprem_transports` guards the invariant.

## What this does NOT validate

- **Training convergence.** Fake GS returns random losses. Real Cortex server is still the gate for loss-curve validation.
- **Full verl / SkyRL driver runs.** Those need `verl` / `skyrl` + `arctic-inference[vllm]` + `tensordict` installed on the driver box. See #49 §3 for the runtime pieces.

## Follow-ups (see [#49](https://github.com/Snowflake-AI-Research/Arctic-Platform/issues/49))

Server-side (Neutrino GS), blocking real Cortex training convergence:

- `POST /{job_id}/forward-no-grad` accepting `{args, kwargs, meta, processing, reference_model?}` (DSSST1 chunks in) → `{model_outputs: {logprobs, entropy?}, metrics: {...}}` (DSSST1 out).
- `POST /{job_id}/log-probs` (JSON in / DSSST1 out).
- Grow `metrics: {grad_norm, ppo_kl, pg_clipfrac_lower, pg_loss, kl_loss, kl_coef}` block on `fwd_bwd` / `step` responses.
- Batch-contract validation for pre-tokenized `kwargs={input_ids, attention_mask, position_ids, prompts, responses, response_mask, advantages, old_log_probs, ref_log_prob?}` on a real cluster.

The `fake_cortex_gs.py` shipped in this PR speaks the exact interface the real server needs to match — GS team can run their implementation against `test_cortex_transport_smoke.py` for a red/green.

## Test plan

- [x] `pytest tests/client/ tests/e2e/` locally — 116 pass on a Cortex-shaped dev env (no ray/vllm/arctic_inference/tensordict).
- [x] AST/import sanity: `arctic_platform.rl.client` no longer exposes `ArcticRLHTTPClient` / `ArcticRLRayClient` at module scope; the cortex dispatch path never imports the on-prem transport modules.
- [x] Env-var override: full 6-test matrix of env states asserts the rewrite behaves like a normal config path.
- [x] End-to-end plumbing: 18 smoke tests hit the fake GS through real HTTP, real DSSST1 chunked upload, real request-status polling. Every op the transport calls round-trips.
- [ ] On-prem regression smoke against SkyRL's `run_qwen3_0.6b_gsm8k_grpo_arl.sh` — loss curves match `main` to within numerical noise. Runs unchanged with `backend: local` (default). Requires the on-prem stack installed.
- [ ] Cortex first-run against the same recipe with `ARCTIC_RL_BACKEND=cortex` + a real Cortex endpoint, once the server-side items in #49 §3 land.

## Superseded

- ~[#51](https://github.com/Snowflake-AI-Research/Arctic-Platform/pull/51)~ — closed. The verl adapter swap it proposed is no longer needed; the env-override + dispatch shim in this PR handles both integrations from the Arctic-Platform side.

### CODE REVIEW

**Comment by sfc-gh-kganesan** [comment]
### End-to-end test on a Cortex QA endpoint (CPU-only driver)

Ran the `recipes/rl/skyrl/simple_gsm8k` recipe against a live Cortex QA sub-account with:

- `ARCTIC_RL_BACKEND=cortex` (PAT auth via `CORTEX_HOST` + `CORTEX_PAT_ENV_VAR`)
- Driver on a CPU-only pod (all GPU work delegated to Cortex, as this PR intends)
- Qwen3-0.6B GRPO, KL disabled (`use_kl_loss=false`, `use_kl_in_reward=false`), `n_samples_per_prompt=4`, `train_batch_size=32`, `update_epochs_per_batch=1`

#### What works out of the box
Job spin-up, `save_weights_for_sampler` → `sync_weights` → eval (`generate` + reward scoring) all complete cleanly; wire framing, DSSST1 chunking, and operation polling behave as designed. The env-var-driven backend override (`ARCTIC_RL_BACKEND=cortex`) does redirect through `arctic_platform.rl` without touching SkyRL/Verl adapters — the "zero integration change" claim holds on the client side.

#### Two CPU-only-driver frictions worth guarding
1. `arctic_platform.rl.utils.server_models` eagerly imports `arctic_inference.server.config.ModelConfig`. On a Cortex-only driver `arctic_inference` isn't (and shouldn't need to be) installed — this makes the whole `arctic_platform.rl` import chain fail. Suggest a lazy / `TYPE_CHECKING` import so the utility modules load without the server package on Cortex drivers.
2. `skyrl/train/utils/utils.py::peer_access_supported` unconditionally spins up a Ray placement group requesting `{'CPU':1,'GPU':2}` when `torch.cuda.is_available()` is False, which hangs the autoscaler forever on a CPU-only head. Not this PR's code, but the Cortex path is the first user to actually hit it — worth an `SKYRL_SKIP_P2P_CHECK=1` escape hatch (either in the arctic_rl entrypoint or upstream SkyRL).

#### Convergence blocker (already called out in the "Follow-ups" section)
Training gets through eval + step-0 sync, then dies on step 1 inside `_ArcticDispatch._compute_old_log_probs` → `client.fwd_no_grad` → HTTP 404 because the Cortex GS doesn't yet expose the `forward-no-grad` endpoint. This matches the follow-up list in the PR description exactly — client side is doing its job, purely waiting on server-side op registration.

Worth noting for reviewers: this call is **not** a KL-loss thing (KL is fully off in the recipe, no reference model instantiated). It's the π_old snapshot for the PPO/GRPO ratio, invoked once per training step regardless of KL config.

Once the endpoint lands (or old-log-probs are folded into `forward-backward`'s response metrics), convergence validation can proceed.
